src/extensions/techno/technoext_hooks.cpp

```diff
@@ -0,0 +1,115 @@
+/*******************************************************************************
+ *
+ *  @project       Vinifera
+ *
+ *  @file          TECHNOEXT_HOOKS.CPP
+ *
+ *  @author        CCHyper
+ *
+ *  @brief         Contains the hooks for the extended TechnoClass.
+ *
+ *  @license       Vinifera is free software: you can redistribute it and/or
+ *                 modify it under the terms of the GNU General Public License
+ *                 as published by the Free Software Foundation, either version
+ *                 3 of the License, or (at your option) any later version.
+ *
+ *                 Vinifera is distributed in the hope that it will be
+ *                 useful, but WITHOUT ANY WARRANTY; without even the implied
+ *                 warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+ *                 PURPOSE. See the GNU General Public License for more details.
+ *
+ *                 You should have received a copy of the GNU General Public
+ *                 License along with this program.
+ *                 If not, see <http://www.gnu.org/licenses/>.
+ *
+ ******************************************************************************/
+#include "technoext_hooks.h"
+#include "tibsun_defines.h"
+#include "techno.h"
+#include "house.h"
+#include "warheadtype.h"
+#include "warheadtypeext.h"
+#include "fatal.h"
+#include "debughandler.h"
+#include "asserthandler.h"
+
+#include "hooker.h"
+#include "hooker_macros.h"
+
+
+/**
+ *  #issue-411
+ * 
+ *  Implements IsAffectsAllies for WarheadTypes.
+ * 
+ *  @note: This patch does not replace "stolen" code as per our implementation
+ *         rules, this is because the call to ObjectClass::Take_Damage that follows
+ *         is too much of a risk to not have correctly implemented.
+ * 
+ *  @author: CCHyper
+ */
+DECLARE_PATCH(_TechnoClass_Take_Damage_IsAffectsAllies_Patch)
+{
+    GET_REGISTER_STATIC(TechnoClass *, this_ptr, esi);
+    GET_STACK_STATIC(int *, damage, esp, 0xEC);
+    GET_STACK_STATIC(int, distance, esp, 0xF0);
+    GET_STACK_STATIC(const WarheadTypeClass *, warhead, esp, 0xF4);
+    GET_STACK_STATIC(TechnoClass *, source, esp, 0xF8);
+    GET_STACK_STATIC8(bool, forced, esp, 0xFC);
+    GET_STACK_STATIC(int, a6, esp, 0x100);
+    static WarheadTypeClassExtension *warheadtypeext;
+    static ResultType result;
+
+    if (warhead) {
+
+        /**
+         *  Is the warhead that hit us one that affects units allied with its firing owner?
+         */
+        warheadtypeext = WarheadTypeClassExtensions.find(warhead);
+        if (warheadtypeext && !warheadtypeext->IsAffectsAllies) {
+
+            /**
+             *  If the source of the damage is an ally of ours, then reset
+             *  the damage amount and return that we took no damage.
+             */
+            if (source && source->House->Is_Ally(this_ptr->House)) {
+                *damage = 0;
+                goto return_RESULT_NONE;
+            }
```

Comment:
Could add [`AffectsEnemies` and `AffectsOwner` from Ares](https://ares-developers.github.io/Ares-docs/new/warheads/general.html) as well, should be a few lines of code.